File: src/test/java/dev/jlo/gamemodes/domain/common/MatchLifecycleTest.java
package dev.jlo.gamemodes.domain.common;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class MatchLifecycleTest {
    @Test
    void lifecycleFollowsCommonArenaFlow() {
        MatchLifecycle lifecycle = new MatchLifecycle();

        assertEquals(MatchPhase.DISABLED, lifecycle.getPhase());
        lifecycle.transitionTo(MatchPhase.WAITING);
        lifecycle.transitionTo(MatchPhase.PREPARING);
        lifecycle.transitionTo(MatchPhase.ACTIVE);
        lifecycle.transitionTo(MatchPhase.RESOLVING);
        lifecycle.transitionTo(MatchPhase.CLEANUP);
        lifecycle.transitionTo(MatchPhase.WAITING);
        assertEquals(MatchPhase.WAITING, lifecycle.getPhase());
    }

    @Test
    void illegalTransitionsAreRejected() {
        MatchLifecycle lifecycle = new MatchLifecycle();

        assertThrows(IllegalStateException.class, () ->
                lifecycle.transitionTo(MatchPhase.ACTIVE));
        lifecycle.transitionTo(MatchPhase.WAITING);
        assertThrows(IllegalStateException.class, () ->
                lifecycle.transitionTo(MatchPhase.CLEANUP));
    }
}

File: src/test/kotlin/dev/jlo/gamemodes/domain/common/MatchLifecycleTest.kt
package dev.jlo.gamemodes.domain.common

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class MatchLifecycleTest {
    @Test
    fun `lifecycle follows common arena flow`() {
        val lifecycle = MatchLifecycle()

        assertEquals(MatchPhase.DISABLED, lifecycle.phase)
        lifecycle.transitionTo(MatchPhase.WAITING)
        lifecycle.transitionTo(MatchPhase.PREPARING)
        lifecycle.transitionTo(MatchPhase.ACTIVE)
        lifecycle.transitionTo(MatchPhase.RESOLVING)
        lifecycle.transitionTo(MatchPhase.CLEANUP)
        lifecycle.transitionTo(MatchPhase.WAITING)
        assertEquals(MatchPhase.WAITING, lifecycle.phase)
    }

    @Test
    fun `illegal transitions are rejected`() {
        val lifecycle = MatchLifecycle()

        assertFailsWith<IllegalStateException> {
            lifecycle.transitionTo(MatchPhase.ACTIVE)
        }
        lifecycle.transitionTo(MatchPhase.WAITING)
        assertFailsWith<IllegalStateException> {
            lifecycle.transitionTo(MatchPhase.CLEANUP)
        }
    }
}
